### PR TITLE
Added ability to invert the test filter by prepending a '^' char

### PR DESCRIPTION
**Pull request by slindberg**
The CI tool we use [recently started supporting running concurrent tests](https://codeship.com/features/parallelci), which I really want to take advantage of. The `--filter` option to `ember test` is perfect for running a subset of tests, however the major drawback is that _all_ tests must be grepabble in a mutually exclusive way, which makes it easy for tests to become accidentally excluded from the CI build. A great way to solve this is to use the [`invert` option](https://github.com/mochajs/mocha/pull/481) in mocha (introduced in v1.3) so that one build pipeline can pick up all non-opt-in tests. This is also just useful to be able to do by itself.

Since ember-cli only has plumbing for passing a filter to the `buildTestPageQueryString` hook, the question becomes how to encode the invert flag into the filter (I'm assuming that since test runners don't universally support this behavior, adding an `--invert` option to `ember test` is not viable). The important things I could think of are:

1. Don't conflict with something you actually want to grep for
2. Don't use characters that are interpolated on the command line
3. Use something that is widely accepted as an inversion operation

The best fit I could come up with is prepending the `^` character at the beginning of the filter, since it's unlikely to exist in test description/names, is cli-friendly afaict, and is how you negate regex character classes. The only downside is that it also happens to be how you anchor a regex match to the beginning of the string, possibly giving the false impression that the filter supports regular expressions.

Thoughts?

### CODE REVIEW

**Comment by dgeb** [comment]
Definitely seems useful. However, I'm concerned about getting too clever with the `filter` option override.

> I'm assuming that since test runners don't universally support this behavior, adding an --invert option to ember test is not viable

Would you mind testing that assumption by raising an issue for ember-cli? A more explicit way of passing this option would be preferable.

**Comment by rwjblue** [comment]
I think we should figure out how to allow arbitrary command options, based on the test implementation.  I think this *should* work if we inherit and implement a custom command with our own options.

tldr; I do *not* think ember-cli should have an `--inverse-filter` (or whatever), but I **do** think that ember-cli-mocha should be able to have that option without having to completely re-implement the test command.

**Comment by dgeb** [comment]
@rwjblue I agree completely. We should push for this type of customization in ember-cli in a flexible manner.

**Comment by rwjblue** [comment]
Just spoke with @dgeb about this in IRC, the current plan of attack is that I will try to spike out the custom test option implementation and see what issues I come across.  If it is possible, I will either PR or link to my branch so that others (potentially @slindberg) can pick up the spike and take it to completion.  If it is not possible we will have much more details on what actually needs to change upstream in ember-cli, and I will open an issue.

Stay tuned....

**Comment by slindberg** [comment]
Sounds good, thanks guys for taking the time to think about this.

> tldr; I do not think ember-cli should have an --inverse-filter (or whatever), but I do think that ember-cli-mocha should be able to have that option without having to completely re-implement the test command. If this is not possible, then that is a bug/issue in ember-cli for sure...

So the idea is that addons can extend/replace existing commands? Or just provide hooks to addons to register custom options to certain commands? I'm still way behind on most of ember-cli's internals, but provided some direction, I can definitely fill out a feature with tests/docs.

**Comment by blimmer** [comment]
Thanks for this @slindberg , I have a need to invert tests, too. For now, I'm forking this behavior. Is there something coming down the line for this kind of support in ember-cli? If not, I'd love to see an interim solution (maybe your [solution](https://github.com/slindberg/ember-cli-mocha/commit/3ec9f9d8aad91b4715192b8ddab33463d1643cbc)) integrated so I don't have to use a fork.

**Comment by slindberg** [comment]
I had a conversation about this with @rwjblue today, and he pointed me to [this ember-cli RFC](https://github.com/davewasmer/rfcs/blob/master/active/0000-extensible-commands.md), which sounds like it has a good chance of being realized. It would allow an addon to extend and add options to the build-in `test` command. In light of this, I think this addon could just temporarily monkey patch the test command's `availableOptions` to add an `invert` option. I don't think this is really a stretch considering this addon is [already modifying the built-in test command's prototype](https://github.com/switchfly/ember-cli-mocha/blob/v0.6.0/index.js#L13).

**Comment by blimmer** [comment]
@slindberg I implemented the monkey-patch in #48 .

**Comment by rwjblue** [comment]
I think this is resolved now. @slindberg can you confirm?

**Comment by slindberg** [comment]
Just tested it out in a dummy app, works perfect. Thanks @blimmer!
